File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid.Core/Component/ControlYesNo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.V4.Content;
using Android.Util;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Class;
using BPMOPMobile.Droid.Core.Class;
using BPMOPMobile.Droid.Core.Common;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Core.Component
{
    public class ControlYesNo : ControlBase
    {
        public LinearLayout _parentView { get; set; }
        public ViewElement _element { get; set; }

        private Switch _switch;

        // Input grid Details 
        public ViewElement _elementParent { get; set; } // Element của Control Grid
        public ViewElement _elementChild { get; set; } // Element child của Item dc click vào
        public int _flagView { get; set; }
        public JObject _JObjectChild { get; set; }

        public override string Value
        {
            set
            {
                if (_element.Enable) // cho chỉnh sửa -> hiện Switch
                {
                    if (value.ToLowerInvariant() == "true")
                        this._switch.Checked = true;
                    else
                        this._switch.Checked = false;
                }
                else // hiện lên _tvValue
                {
                    if (value.ToLowerInvariant() == "true")
                        this._tvValue.Text = CmmFunction.GetTitle("TEXT_LABEL_YES", "Có");
                    else
                        this._tvValue.Text = CmmFunction.GetTitle("TEXT_LABEL_NO", "Không");
                }
            }
            get { return this._switch.Text.ToString(); }
        }

        public ControlYesNo(Activity _mainAct, LinearLayout _parentView, ViewElement _element) : base(_mainAct)
        {
            this._parentView = _parentView;
            this._element = _element;
            InitializeComponent();
        }
        /// <summary>
        /// Constructor của Input Grid Details
        /// </summary>
        public ControlYesNo(Activity _mainAct, LinearLayout _parentView, ViewElement _elementParent, ViewElement _elementChild, JObject _JObjectChild, int _flagView) : base(_mainAct)
        {
            // Để render theo base
            this._parentView = _parentView;
            this._element = _elementChild;

            // Input grid Details
            this._elementParent = _elementParent;
            this._elementChild = _elementChild;
            this._JObjectChild = _JObjectChild;
            this._flagView = _flagView;
            InitializeComponent();
        }

        public override void InitializeComponent()
        {
            base.InitializeComponent();

            _switch = new Switch(_mainAct);

            //if (_lnContent != null)
            //{
            //    _lnContent.Click += HandleTouchDown;
            //}
        }
        public override void InitializeFrameView(LinearLayout frame)
        {
            if (_element.Hidden == true) // Check xem có ẩn view hay không
                return;

            base.InitializeFrameView(frame);

            if (_element.Enable == true) // cho chỉnh sửa -> hiện Switch
            {
                this._tvValue.Visibility = ViewStates.Gone;

                int _padding = (int)TypedValue.ApplyDimension(ComplexUnitType.Dip, 6, _mainAct.Resources.DisplayMetrics);
                LinearLayout.LayoutParams _switchParams = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WrapContent, LinearLayout.LayoutParams.WrapContent);
                _switch.SetPadding(0, _padding, _padding, _padding);
                _switch.LayoutParameters = _switchParams;

                _switch.CheckedChange += HandleTouchDown;

                _lnContent.AddView(_switch);
            }
            else // hiện lên _tvValue
            {
                this._tvValue.Visibility = ViewStates.Visible;
            }
        }

        /// <summary>
        /// Event khi click vào Control
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        protected virtual void HandleTouchDown(object sender, CompoundButton.CheckedChangeEventArgs e)
        {
            if (e.IsChecked == true)
                _element.Value = "True";
            else
                _element.Value = "False";

            if (_parentView != null)
            {
                if (_JObjectChild != null) // Input grid Detail
                {
                    MinionActionCore.OnElementGridChildActionClick(null, new MinionActionCore.ElementGridChildActionClick(_elementParent, _elementChild, _JObjectChild, _flagView));
                }
                else
                {
                    MinionActionCore.OnElementFormClick(null, new MinionActionCore.ElementFormClick(_element));
                }
            }
        }
        public override void SetProprety()
        {
            if (_element.ListProprety != null)
            {
                foreach (var item in _element.ListProprety)
                {
                    CmmDroidFunction.SetPropertyValueByNameCustom(_tvValue, item.Key, item.Value);
                }
            }
        }
        public override void SetEnable()
        {
            base.SetEnable();

            if (_element.Enable)
            {
                _switch.Clickable = true;
                _parentView.Clickable = true;
            }
            else
            {
                _switch.Clickable = false;
                _parentView.Clickable = false;
            }
        }
        public override void SetTitle()
        {
            base.SetTitle();

            _tvTitle.Text = _element.Title;

            if (_element.IsRequire && _element.Enable)
            {
                _tvTitle.Text += " (*)";
                CmmDroidFunction.SetTextViewHighlightControl(_mainAct, _tvTitle);
            }
        }
        public override void SetValue()
        {
            base.SetValue();

            if (_element.Enable == true)
                _switch.CheckedChange -= HandleTouchDown; // để khỏi trigger event

            Value = _element.Value;

            if (_element.Enable == true)
                _switch.CheckedChange += HandleTouchDown;

        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/core/component/ControlYesNo.java
package com.vuthao.bpmop.core.component;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.util.TypedValue;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.Switch;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;

import org.json.JSONObject;

public class ControlYesNo extends ControlBase implements CompoundButton.OnCheckedChangeListener {
    private final LinearLayout parentView;
    private final ViewElement element;
    private Switch switchYesNo;
    // Input grid Details
    private ViewElement elementParent; // Element của Control Grid
    private ViewElement elementChild; // Element child của Item dc click vào
    private int flagView;
    private JSONObject JObjectChild;

    public ControlYesNo(Activity mainAct, LinearLayout parentView, ViewElement element) {
        super(mainAct);
        this.parentView = parentView;
        this.element = element;
        initializeComponent();
    }

    /// <summary>
    /// Constructor của Input Grid Details
    /// </summary>
    public ControlYesNo(Activity mainAct, LinearLayout parentView, ViewElement elementParent, ViewElement elementChild, JSONObject JObjectChild, int flagView) {
        super(mainAct);
        // Để render theo base
        this.parentView = parentView;
        this.element = elementChild;

        // Input grid Details
        this.elementParent = elementParent;
        this.elementChild = elementChild;
        this.JObjectChild = JObjectChild;
        this.flagView = flagView;
        initializeComponent();
    }

    @Override
    public void setValue(String value) {
        if (element.isEnable()) {
            switchYesNo.setChecked(value.toLowerCase().equals("true"));
        } else {
            if (value.toLowerCase().equals("true")) {
                tvValue.setText(Functions.share.getTitle("TEXT_LABEL_YES", "Có"));
            } else {
                tvValue.setText(Functions.share.getTitle("TEXT_LABEL_NO", "Không"));
            }
        }
    }

    @Override
    public String getValue() {
        return switchYesNo.getText().toString();
    }

    @Override
    public void initializeComponent() {
        super.initializeComponent();

        switchYesNo = new Switch(mainAct);
    }

    @Override
    public void initializeFrameView(LinearLayout frame) {
        if (element.isHidden()) return;
        super.initializeFrameView(frame);

        if (element.isEnable()) {
            tvValue.setVisibility(View.GONE);

            int padding = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6, mainAct.getResources().getDisplayMetrics());
            LinearLayout.LayoutParams _switchParams = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
            switchYesNo.setPadding(0, padding, padding, padding);
            switchYesNo.setLayoutParams(_switchParams);

            switchYesNo.setOnCheckedChangeListener(this);

            lnContent.addView(switchYesNo);
        } else {
            tvValue.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        if (isChecked) {
            element.setValue("True");
        } else {
            element.setValue("False");
        }

        if (parentView != null) {
            Intent intent = new Intent();
            if (JObjectChild != null) {
                intent.setAction(VarsReceiver.CHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementChild));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("flagview", flagView);
            } else {
                intent.setAction("FORMCLICK");
                intent.putExtra("element", new Gson().toJson(element));
            }
            BroadcastUtility.send(mainAct, intent);
        }
    }

    @Override
    public void setProprety() {
        super.setProprety();

    }

    @Override
    public void setEnable() {
        super.setEnable();

        if (element.isEnable()) {
            switchYesNo.setClickable(true);
            parentView.setClickable(true);
        } else {
            switchYesNo.setClickable(false);
            parentView.setClickable(false);
        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void setTitle() {
        super.setTitle();

        tvTitle.setText(element.getTitle());
        if (element.isRequire() && element.isEnable()) {
            tvTitle.setText(tvTitle.getText() + " (*)");
            Functions.share.setTVHighlightControl(mainAct, tvTitle);
        }
    }

    @Override
    public void setValue() {
        super.setValue();

        if (element.isEnable()) {
            switchYesNo.setOnCheckedChangeListener(null);
        }
        setValue(element.getValue());

        if (element.isEnable()) {
            switchYesNo.setOnCheckedChangeListener(this);
        }
    }
}
